risc0/zkvm/platform/src/syscall.rs

```diff
@@ -141,85 +146,211 @@ impl SyscallName {
+    #[cfg(target_os = "zkvm")]
+    {
+        let a0: u32;
+        let a1: u32;
+        asm!(
+            "ecall",
+            in("t0") ecall::SOFTWARE,
+            inout("a0") from_host => a0,
+            inout("a1") from_host_words => a1,
+            in("a2") syscall.as_ptr(),
+            in("a3") a3,
+            in("a4") a4,
+        );
+        Return(a0, a1)
+    }
+    #[cfg(not(target_os = "zkvm"))]
+    unimplemented!()
+}
+
+/// Invoke a raw system call
+#[no_mangle]
+pub unsafe extern "C" fn syscall_3(
+    syscall: SyscallName,
+    from_host: *mut u32,
+    from_host_words: usize,
+    a3: u32,
+    a4: u32,
+    a5: u32,
+) -> Return {
+    #[cfg(target_os = "zkvm")]
+    {
+        let a0: u32;
+        let a1: u32;
+        asm!(
+            "ecall",
+            in("t0") ecall::SOFTWARE,
+            inout("a0") from_host => a0,
+            inout("a1") from_host_words => a1,
+            in("a2") syscall.as_ptr(),
+            in("a3") a3,
+            in("a4") a4,
+            in("a5") a5,
+        );
+        Return(a0, a1)
     }
+    #[cfg(not(target_os = "zkvm"))]
+    unimplemented!()
 }
 
-impl_syscall!(syscall_0);
-impl_syscall!(syscall_1, a3);
-impl_syscall!(syscall_2, a3, a4);
-impl_syscall!(syscall_3, a3, a4, a5);
-impl_syscall!(syscall_4, a3, a4, a5, a6);
-impl_syscall!(syscall_5, a3, a4, a5, a6, a7);
+/// Invoke a raw system call
+#[no_mangle]
+pub unsafe extern "C" fn syscall_4(
+    syscall: SyscallName,
+    from_host: *mut u32,
+    from_host_words: usize,
+    a3: u32,
+    a4: u32,
+    a5: u32,
+    a6: u32,
+) -> Return {
+    #[cfg(target_os = "zkvm")]
+    {
+        let a0: u32;
+        let a1: u32;
+        asm!(
+            "ecall",
+            in("t0") ecall::SOFTWARE,
+            inout("a0") from_host => a0,
+            inout("a1") from_host_words => a1,
+            in("a2") syscall.as_ptr(),
+            in("a3") a3,
+            in("a4") a4,
+            in("a5") a5,
+            in("a6") a6,
+        );
+        Return(a0, a1)
+    }
+    #[cfg(not(target_os = "zkvm"))]
+    unimplemented!()
+}
+
+/// Invoke a raw system call
+#[no_mangle]
+pub unsafe extern "C" fn syscall_5(
+    syscall: SyscallName,
+    from_host: *mut u32,
+    from_host_words: usize,
+    a3: u32,
+    a4: u32,
+    a5: u32,
+    a6: u32,
+    a7: u32,
+) -> Return {
+    #[cfg(target_os = "zkvm")]
+    {
+        let a0: u32;
+        let a1: u32;
+        asm!(
+            "ecall",
+            in("t0") ecall::SOFTWARE,
+            inout("a0") from_host => a0,
+            inout("a1") from_host_words => a1,
+            in("a2") syscall.as_ptr(),
+            in("a3") a3,
+            in("a4") a4,
+            in("a5") a5,
+            in("a6") a6,
+            in("a7") a7,
+        );
+        Return(a0, a1)
+    }
+    #[cfg(not(target_os = "zkvm"))]
+    unimplemented!()
+}
 
 #[inline(always)]
 #[no_mangle]
-pub unsafe extern "C" fn sys_halt() {
+pub unsafe extern "C" fn sys_halt() -> ! {
```

Comment:
Could we have sys_halt take a halt::* argument instead of having three separate asm blocks?

Reply:
There's only two cases (AFAICT) that are user-visible (HALT & PAUSE). I guess I left it separate because perhaps the interface for pause will change, and also `sys_halt` wants to return `!` whereas `sys_pause` can't do that.